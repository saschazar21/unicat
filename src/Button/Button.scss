@import '../__styles__/variables';

.wrapper {
  display: inline-flex;
  box-shadow: boxShadow(xs);
  align-items: center;
  justify-content: center;
  flex-wrap: nowrap;
  border: none;
  border-radius: borderRadius();
  padding: spacing(m) spacing(s);
  max-width: 100%;
  font-size: fontSize(small);
  letter-spacing: 0.25em;
  text-transform: uppercase;
  transition: transition(box-shadow);

  &:hover,
  &:focus,
  &:active {
    box-shadow: boxShadow(m);
    cursor: pointer;
  }
}

.block {
  display: flex;
}

.large {
  padding: spacing(l) spacing(m);
  font-size: fontSize(h6);
}

@each $variant in $variants {
  .#{$variant} {
    background-color: color($variant);
    color: color($variant, true);
  }
}

.disabled {
  background-color: transparentize(color(default), 0.75);
  color: color(default);
  cursor: default;

  &:hover,
  &:focus,
  &:active {
    cursor: default;
    box-shadow: boxShadow(xs);
  }
}
