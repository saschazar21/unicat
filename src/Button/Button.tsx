import React, { Component } from 'react';
import classnames from 'classnames';

import styles from './Button.scss';

export interface ButtonProps {
  block?: boolean;
  className?: string;
  disabled?: boolean;
  icon?: SVGSVGElement;
  large?: boolean;
  onClick?: any;
  title: string;
  type?: 'submit' | 'reset';
  variant?: Variant;
}

export default class Button extends Component<ButtonProps> {
  static defaultProps = {
    block: false,
    variant: 'primary',
    type: 'submit',
  };

  public render() {
    const { variant: defaultVariant } = Button.defaultProps;
    const {
      className: customClassName,
      disabled,
      icon,
      large,
      onClick,
      title,
      type,
      variant = defaultVariant,
    } = this.props;

    const className = classnames(
      styles.wrapper,
      styles[variant],
      { [styles.large]: large, [styles.disabled]: disabled },
      customClassName
    );

    return (
      <button className={className} onClick={onClick} type={type}>
        {icon && <span>{icon}</span>}
        {title}
      </button>
    );
  }
}
