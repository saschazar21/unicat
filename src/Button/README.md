Renders a `<button />`.

```jsx
<Button title="A default button" />
```

An alternative `<button />` may be rendered using the `light` variant.

```jsx
<Button title="A light button" variant="light" />
```

If a larger version is wanted, the `large` prop type may be added.

```jsx
<Button title="A large button" large />
```

To disable a button, the `disabled` prop works just like normal.

```jsx
<Button title="A disabled button" disabled />
```
