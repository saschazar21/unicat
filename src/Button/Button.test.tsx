import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import Button from './Button';

describe('Button', () => {
  it('renders', () => {
    const mock = jest.fn();
    const wrapper = shallow(<Button onClick={mock} title="Test" />);

    expect(toJson(wrapper)).toMatchSnapshot();
    expect(wrapper.prop('type')).toEqual('submit');
    expect(mock).not.toHaveBeenCalled();

    wrapper.simulate('click');

    expect(mock).toHaveBeenCalled();
  });

  it('renders a light variant', () => {
    const mock = jest.fn();
    const wrapper = shallow(<Button onClick={mock} title="Test" variant="light" />);

    expect(toJson(wrapper)).toMatchSnapshot();

    expect(wrapper.find('.light')).toHaveLength(1);
    expect(wrapper.find('.default')).toHaveLength(0);
  });

  it('renders a disabled variant', () => {
    const mock = jest.fn();
    const wrapper = shallow(<Button onClick={mock} title="Test" disabled />);

    expect(toJson(wrapper)).toMatchSnapshot();
    expect(mock).not.toHaveBeenCalled();

    wrapper.simulate('click');

    expect(mock).not.toHaveBeenCalled();
  });
});
