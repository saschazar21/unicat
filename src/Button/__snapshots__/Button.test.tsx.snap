// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button renders 1`] = `
<button
  className="wrapper default"
  onClick={[MockFunction]}
  type="submit"
>
  Test
</button>
`;

exports[`Button renders a disabled variant 1`] = `
<button
  className="wrapper disabled"
  disabled={true}
  type="submit"
>
  Test
</button>
`;

exports[`Button renders a light variant 1`] = `
<button
  className="wrapper light"
  onClick={[MockFunction]}
  type="submit"
>
  Test
</button>
`;
